core/src/block_strider/archive_handler.rs

```diff
@@ -0,0 +1,129 @@
+use tokio::sync::watch;
+use tycho_storage::Storage;
+use tycho_util::metrics::HistogramGuard;
+
+use crate::block_strider::{
+    ArchiveSubscriber, ArchiveSubscriberContext, BlockSubscriber, BlockSubscriberContext,
+};
+
+#[repr(transparent)]
+pub struct ArchiveHandler<S> {
+    inner: Arc<Inner<S>>,
+}
+
+impl<S> ArchiveHandler<S>
+where
+    S: ArchiveSubscriber,
+{
+    pub fn new(storage: Storage, archive_subscriber: S) -> Self {
+        let archive_listener = storage.block_storage().archive_listener().clone();
+        Self {
+            inner: Arc::new(Inner {
+                storage,
+                archive_subscriber,
+                archive_listener: ArchiveListener::new(archive_listener),
+            }),
+        }
+    }
+
+    async fn prepare_block_impl(
+        &self,
+        _cx: &BlockSubscriberContext,
+    ) -> Result<ArchiveHandlerPrepared> {
+        let _histogram = HistogramGuard::begin("tycho_core_archive_handler_prepare_block_time");
+
+        Ok(ArchiveHandlerPrepared {})
+    }
+
+    async fn handle_block_impl(
+        &self,
+        cx: &BlockSubscriberContext,
+        _prepared: ArchiveHandlerPrepared,
+    ) -> Result<()> {
+        let _histogram = HistogramGuard::begin("tycho_core_archive_handler_handle_block_time");
+
+        tracing::info!(id = %cx.block.id(), "handling block");
+
+        // Process archive
+        if let Some(archive_id) = self.inner.archive_listener.next_update()? {
+            let started_at = std::time::Instant::now();
+
+            let cx = ArchiveSubscriberContext {
+                archive_id,
+                storage: self.inner.storage.clone(),
+            };
+
+            self.inner.archive_subscriber.handle_archive(&cx).await?;
+
+            metrics::histogram!("tycho_core_subscriber_handle_archive_time")
+                .record(started_at.elapsed());
+        }
+
+        // Done
+        Ok(())
+    }
+}
+
+impl<S> Clone for ArchiveHandler<S> {
+    #[inline]
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+impl<S> BlockSubscriber for ArchiveHandler<S>
+where
+    S: ArchiveSubscriber,
+{
+    type Prepared = ArchiveHandlerPrepared;
+
+    type PrepareBlockFut<'a> = BoxFuture<'a, Result<Self::Prepared>>;
+    type HandleBlockFut<'a> = BoxFuture<'a, Result<()>>;
+
+    fn prepare_block<'a>(&'a self, cx: &'a BlockSubscriberContext) -> Self::PrepareBlockFut<'a> {
+        Box::pin(self.prepare_block_impl(cx))
+    }
+
+    fn handle_block<'a>(
+        &'a self,
+        cx: &'a BlockSubscriberContext,
+        prepared: Self::Prepared,
+    ) -> Self::HandleBlockFut<'a> {
+        Box::pin(self.handle_block_impl(cx, prepared))
+    }
+}
+
+pub struct ArchiveHandlerPrepared {}
+
+struct Inner<S> {
+    storage: Storage,
+    archive_subscriber: S,
+    archive_listener: ArchiveListener,
+}
+
+struct ArchiveListener {
+    rx: parking_lot::Mutex<watch::Receiver<u32>>,
```

Comment:
Just clone receiver for borrow and update i. It's an Arc + Version.

Reply:
fixed